File: AOC_2022/Day 8/8.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <cstring>
#include <cctype>
#include <vector>
#include <algorithm>
using namespace std;

int main()
{
    ifstream in("input.txt");
    if (!in.is_open())
    {
        cerr << "Could not open the file - '" << endl;
        return EXIT_FAILURE;
    }
    string line;
    string grid[99];
    int p = 0;
    while (getline(in, line))
    {
        grid[p++] = line;
    }
    bool visible[99][99];
    for (int i = 0; i < 99; i++)
    {
        for (int j = 0; j < 99; j++)
        {
            if (i == 0 || i == 98 || j == 0 || j == 98)
            {
                visible[i][j] = true;
            }
            else
                visible[i][j] = false;
        }
    }
    int max_height;
    for (int i = 0; i < 99; i++)
    {
        max_height = 0;
        for (int j = 0; j < 99; j++)
        {
            if (grid[i][j] > max_height)
            {
                max_height = grid[i][j];
                visible[i][j] = true;
            }
        }
        max_height = 0;
        for (int j = 98; j >= 0; j--)
        {
            if (grid[i][j] > max_height)
            {
                max_height = grid[i][j];
                visible[i][j] = true;
            }
        }
    }
    for (int j = 0; j < 99; j++)
    {
        max_height = 0;
        for (int i = 0; i < 99; i++)
        {
            if (grid[i][j] > max_height)
            {
                max_height = grid[i][j];
                visible[i][j] = true;
            }
        }
        max_height = 0;
        for (int i = 98; i >= 0; i--)
        {
            if (grid[i][j] > max_height)
            {
                max_height = grid[i][j];
                visible[i][j] = true;
            }
        }
    }
    int result = 0;
    for (int i = 0; i < 99; i++)
    {
        for (int j = 0; j < 99; j++)
        {
            if (visible[i][j])
            {
                result++;
            }
        }
    }
    cout << result << "\n";

    // part 2
    result = 0;
    for(int i = 1; i < 98;i++){
        for(int j = 1; j < 98;j++){
            int curr = 1,k;
            for(k = i+1; k < 98;k++){
                if(grid[i][j] <= grid[k][j])
                    break;
            }
            curr *= (k-i);
            for(k = i-1; k > 0;k--){
                if(grid[i][j] <= grid[k][j])
                    break;
            }
            curr *= (i-k);
            for(k = j+1; k < 98;k++){
                if(grid[i][j] <= grid[i][k])
                    break;
            }
            curr *= (k-j);
            for(k = j-1; k > 0;k--){
                if(grid[i][j] <= grid[i][k])
                    break;
            }
            curr *= (j-k);
            result = max(result,curr);
        }
    }
    cout << result << "\n";
    return 0;
}

File: AOC_2022/Day 8/8.py
file = open("input.txt", "r")

lines = file.read().splitlines()

grid = [[False for _ in range(len(lines[0]))] for _ in range(len(lines))]

for i in range(len(lines)):
    grid[i][0] = True
    grid[i][len(lines) - 1] = True

for i in range(len(lines[0])):
    grid[0][i] = True
    grid[len(lines[0]) - 1][i] = True

for i in range(len(lines)):
    maximum = 0
    for j in range(len(lines[i])):
        if int(lines[i][j]) > maximum :
            grid[i][j] = True
            maximum = int(lines[i][j])
    maximum = 0
    for j in range(len(lines[0])-1,0,-1):
        if int(lines[i][j]) > maximum :
            grid[i][j] = True
            maximum = int(lines[i][j])

for i in range(len(lines[0])):
    maximum = 0
    for j in range(len(lines)):
        if int(lines[j][i]) > maximum :
            grid[j][i] = True
            maximum = int(lines[j][i])
    maximum = 0
    for j in range(len(lines)-1,0,-1):
        if int(lines[j][i]) > maximum :
            grid[j][i] = True
            maximum = int(lines[j][i])

result = sum([1 for row in grid for value in row if value])

print("The Result for the 1st star is: " + str(result))

result = 0

for i in range(1,len(lines)-1):
    for j in range(1,len(lines[i])-1):
        curr = 1
        k = j+1
        while(k < len(lines[i])-1 and lines[i][k] < lines[i][j]):
            k += 1
        curr *= k-j
        k = j-1
        while(k > 0 and lines[i][k] < lines[i][j]):
            k -= 1
        curr *= j-k
        k = i+1
        while(k < len(lines)-1 and lines[k][j] < lines[i][j]):
            k += 1
        curr *= k-i
        k = i-1
        while(k > 0 and lines[k][j] < lines[i][j]):
            k -= 1
        curr *= i-k
        result = max(result, curr)

print("The Result for the 2nd star is: " + str(result))